File: problem_solving/math/py/66_plus_one.py
'''
https://leetcode.com/problems/plus-one

start traversing from right..
if last digit is 9, make it 0
if last digit is not 9, just add +1 and return
    at end add 1 to array.. case of all 999
'''

def plus_one(input: list[int]) -> list[int]:
    N=len(input)
    for i in range(N):
        idx = N-1-i 
        if input[idx] == 9:
            input[idx] = 0
        else:
            input[idx] += 1
            return input
    return [1] + input





digits = [1,2,3]
#Output: [1,2,4]
output = plus_one(digits)
print("output: " + str(output))

digits = [4,3,2,1]
#Output: [4,3,2,2]
output = plus_one(digits)
print("output: " + str(output))

File: problem_solving/math/cpp/66_plus_one.cpp
/*
https://leetcode.com/problems/plus-one/

iterate from end.. till carry = 0
*/

#include<iostream>
#include<vector>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"

std::vector<int> plus_one(std::vector<int>& nums) {
    int carry = 1;
    int N=nums.size();
    for(int i = N-1; i >= 0; --i ) {
        nums[i] += carry;
        carry = nums[i] / 10; 
        nums[i] %= 10;
        if(carry == 0) {
            return nums;
        }
    }
    if(carry > 0) {
        nums.insert(nums.begin(), carry);
    }
    
    return nums;
}

std::vector<int> plus_one_new_number(std::vector<int>& nums) {
    int carry = 1;
    int N=nums.size();
    std::vector<int> result=nums;
    for(int i = N-1; i >= 0; --i ) {
        result[i] += carry;
        carry = result[i] / 10; 
        result[i] %= 10;
        if(carry == 0) {
            return result;
        }
    }
    if(carry > 0) {
        result.insert(nums.begin(), carry);
    }
    
    return result;
}



int main() {
    std::vector<int> input = {1,2,3};
    print_vector(input);
    auto result = plus_one(input);
    print_vector(result);

    input = {1,2,3};
    result = plus_one_new_number(input);
    print_vector(result);
}